Xamarin/FormsClient/FormsClient/MainPage.xaml.cs

```diff
@@ -0,0 +1,147 @@
+	{
+        private ViewState _state = ViewState.Disconnected;
+
+	    private ObservableCollection<string> _messages = new ObservableCollection<string>();
+        private HubConnection _connection;
+
+		public MainPage()
+		{
+			InitializeComponent();
+		    MessagesListView.ItemsSource = _messages;
+
+		}
+
+        private async void ConnectDisconnectButton_Clicked(object sender, EventArgs e)
+        {
+            if (_state == ViewState.Connected)
+            {
+                UpdateState(ViewState.Disconnecting);
+
+                try
+                {
+                    await _connection.StopAsync();
+                }
+                catch (Exception ex)
+                {
+                    AppendMessage($"An error occurred while disconnecting: {ex}");
+                    UpdateState(ViewState.Connected);
+                    return;
+                }
+
+                UpdateState(ViewState.Disconnected);
+            }
+            else
+            {
+                UpdateState(ViewState.Connecting);
+
+                try
+                {
+                    _connection = new HubConnectionBuilder()
+                        .WithUrl(ServerUrlEntry.Text)
+                        .Build();
+
+                    _connection.On<string>("Send", (message) =>
+                    {
+                        AppendMessage(message);
+                    });
+
+                    await _connection.StartAsync();
+                }
+                catch (Exception ex)
+                {
+                    AppendMessage($"An error occurred while connecting: {ex}");
+                    UpdateState(ViewState.Disconnected);
+                    return;
+                }
+
+                UpdateState(ViewState.Connected);
+            }
+        }
+
+        private async void SendButton_Clicked(object sender, EventArgs e)
+        {
+            if (_state != ViewState.Connected)
+            {
+                await DisplayAlert("Error", "Must be Connected to Send!", "OK");
+                return;
+            }
+
+            try
+            {
+                await _connection.InvokeAsync("Send", MessageEntry.Text);
+                MessageEntry.Text = "";
+            }
+            catch (Exception ex)
+            {
+                AppendMessage($"An error occurred while sending: {ex}");
+            }
+        }
+
+        private void AppendMessage(string message)
+        {
+            Device.BeginInvokeOnMainThread(() =>
+            {
+                _messages.Add(message);
+                MessagesListView.ScrollTo(_messages.Last(), ScrollToPosition.End, animated: true);
+            });
+        }
+
+        private void UpdateState(ViewState state)
+        {
+            if (_state == state)
+            {
+                return;
+            }
+
+            switch (state)
+            {
+                case ViewState.Disconnected:
+                    SendButton.IsEnabled = false;
```

Comment:
Similar comment about reducing logic in each of the cases by checking specifically for connected. Might not be worth it tbh